server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistTestsStep.java

```diff
@@ -0,0 +1,193 @@
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+public class PersistTestsStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final System2 system;
+
+  public PersistTestsStep(DbClient dbClient, System2 system) {
+    this.dbClient = dbClient;
+    this.system = system;
+  }
+
+  @Override
+  public String[] supportedProjectQualifiers() {
+    return new String[] {Qualifiers.PROJECT};
+  }
+
+  @Override
+  public void execute(ComputationContext computationContext) {
+    DbSession session = dbClient.openSession(false);
+    try {
+      int rootComponentRef = computationContext.getReportMetadata().getRootComponentRef();
+      TestContext context = new TestContext(session, computationContext);
+      recursivelyProcessComponent(context, rootComponentRef);
+      session.commit();
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+  }
+
+  private void recursivelyProcessComponent(TestContext context, int componentRef) {
+    BatchReportReader reportReader = context.reader;
+    BatchReport.Component component = reportReader.readComponent(componentRef);
+    if (component.getIsTest() && reportReader.readTests(componentRef) != null) {
+      persistTestResults(component, context);
+    }
+
+    for (Integer childRef : component.getChildRefList()) {
+      recursivelyProcessComponent(context, childRef);
+    }
+  }
+
+  private void persistTestResults(BatchReport.Component component, TestContext context) {
+    ReportIterator<BatchReport.Test> testIterator = new ReportIterator<>(context.reader.readTests(component.getRef()), BatchReport.Test.PARSER);
+    ListMultimap<String, FileSourceDb.Test.CoveredFile> nameToCoveredFiles = loadCoverageDetails(component.getRef(), context);
+    List<FileSourceDb.Test> tests = new ArrayList<>();
+
+    while (testIterator.hasNext()) {
+      BatchReport.Test batchTest = testIterator.next();
+      FileSourceDb.Test.Builder dbTest = FileSourceDb.Test.newBuilder();
+      dbTest.setFileUuid(component.getUuid());
+      dbTest.setType(batchTest.getType());
+      dbTest.setName(batchTest.getName());
+      if (batchTest.hasStacktrace()) {
+        dbTest.setStacktrace(batchTest.getStacktrace());
+      }
+      if (batchTest.hasStatus()) {
+        dbTest.setStatus(batchTest.getStatus());
+      }
+      if (batchTest.hasMsg()) {
+        dbTest.setMsg(batchTest.getMsg());
+      }
+      if (batchTest.hasExecutionTimeMs()) {
+        dbTest.setExecutionTimeMs(batchTest.getExecutionTimeMs());
+      }
+      List<FileSourceDb.Test.CoveredFile> coveredFiles = nameToCoveredFiles == null ? null : nameToCoveredFiles.get(batchTest.getName());
+      if (coveredFiles != null) {
+        dbTest.addAllCoveredFile(coveredFiles);
+      }
+      tests.add(dbTest.build());
+    }
+
+    FileSourceDto existingDto = dbClient.fileSourceDao().selectTest(component.getUuid());
```

Comment:
Aie !!! You can't execute one query per test, it will take too much time ! You have to use the same strategy as PersistFileSourcesStep -> Do one query for the whole project that returns only uuids of test files.